### ISSUE TITLE
CSVY wishlist

### ISSUE THREAD

**Issue by MichaelChirico** [Rdatatable/data.table#3540]
Filing a follow-up to #1701 & #3534 which handled basic implementation of CSVY support to centralize a wish list for more advanced functionality for CSVY:

### Both fread and fwrite

- [ ] Support for writing a factor's levels & then reading them
- [ ] `yaml = '/path/to/file'` -- store/read metadata stored outside the file itself
- [ ] `yaml = 'auto'` by default -- read/write csvy files automatically by detecting the `csvy` extension

### `fread`

- [ ] Fully sync `fread` arguments with what's output from `fwrite` by default
- [ ] Expand recognized `colClasses` following #2545 
- [ ] auto yaml detection using file contents even if extension is not `.csvy`

### `fwrite`

- [ ] `yaml = list()` -- supply custom key/value pairs to be included in the metadata

**Comment by rossholmberg** [comment]
I'd love to see more user-oriented metadata. Description/note options would be fantastic for storing something along the lines of a brief data readme. I played with something like this myself a while ago (it's all on GitHub), and used an object attribute to store these things in the R object. Here's a section of my readme showing what I mean:
https://github.com/rossholmberg/rcsv/blob/master/README.md#notes-an-added-detail-to-your-datasets

I only had "notes" for the table as a whole, but per-column would be even better.

**Comment by MichaelChirico** [comment]
as in, how to provide customized metadata right? if so yes!! (if not please correct me 😁) something I've thought about but I'm not sure the best API for enabling this. could be that we force user to provide an `attribute` but that feels inconvenient.

**Comment by jangorecki** [comment]
I would lean towards generalising to store DT/cols attributes in yaml metadata. Then we don't need to worry if it is a column attribute, or a table attribute, what is its purpose. This way we handle factor levels, datetime origin/timezone etc. with column attributes, and sorted attribute of a data.table.
Handling indexes is possible (adding technical `__index_cols` integer columns) but probably should be disabled by default it costs extra space, read/write time, and at the same time it is something that can be easily and pretty fast computed.

**Comment by MichaelChirico** [comment]
Makes sense. Agree we shouldn't be writing the index vector, but it would be useful to have a `sortedBy` attribute (or similar) that can indicate the data is already sorted by some columns.

**Comment by rossholmberg** [comment]
I agree Jan, that would be great to be able to store attributes at both DT and col levels.
Yes Michael, customised metadata is exactly what I mean. A few use cases are data source attribution, method notes, measurement units, notes about equipment used, explanations of coded variables, etc. I use the description attributes in SQL a lot for these things, and I find having them stored alongside the data itself to be more tidy than in a separate metadata file.

**Comment by eliocamp** [comment]
I've recently stumbled upon the need to store custom fields in csvy metadata and fully support this. I think something like a `yaml.exta` parameter which accepts a list to be appended to the current automatically-generated list would make sense.

**Comment by eliocamp** [comment]
This code works as advertised. 

https://github.com/eliocamp/data.table/commit/6d4df053d05004da696b28668b963be092e27b86

``` r
library(data.table)

data <- data.table(x = 1, y = 100)
file <- tempfile()

fwrite(data, file, yaml =  list(something_else = "o hai!"))

x <- fread(file, yaml = TRUE)
attr(x, "yaml_metadata")$something_else
#> [1] "o hai!"
```

<sup>Created on 2019-12-16 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

I can start a PR, but I would need some guidance to write tests for this functionality, since I don't understand how tests work in data.table.

**Comment by MichaelChirico** [comment]
@eliocamp that would be great! Something like that is on my radar to add when I get back to this.

Have you seen the section on testing in the Contributing wiki?

https://github.com/Rdatatable/data.table/wiki/Contributing

**Comment by eliocamp** [comment]
Thanks there it is.

**Comment by tdeenes** [comment]
I do not know whether I should open a separate issue for this feature request, so just put it here. 

It would be great if `fread` or a dedicated function built upon the already implemented functionality of `fread` could create an R object or a yaml counterpart representing the metadata of the file (or command, input) argument of `fread`. So instead of running `fread` the user could just call `fglimpse` and get back the number of rows, columns, column header, colClasses, sample of values etc. 

As an intermediate step, it might be enough to be able to get back the metadata that `fread` figures out automagically as an R object (probably as an attribute of the `fread` result) - just as if `verbose = TRUE` returned the results as a list instead of printing it to stdout.

**Comment by jangorecki** [comment]
Sounds like a new request. Note that colClasses might not be accurate, because it can still happens that during full data reading there is type promotion. What fread guess at the beginning is only a guess, not confirmed by full file read.

**Comment by eliocamp** [comment]
Re: "Support for writing a factor's levels & then reading them"

Wouldn't that be a more general issue of preserving all column attributes?

**Comment by MichaelChirico** [comment]
I think yes, but `factor` and `Date`/`POSIXct` will definitely be the bulk of use cases for that

**Comment by eliocamp** [comment]
The previous PR would solve the issue of factor levels and other column attributes. However, while column class is saved (only the first one), they are not recovered by `fread(..., yaml = TRUE)` 

``` r
library(data.table)
file <- tempfile()
data <- data.table(x = 1, y = 100, 
                  c = factor(letters[1:3], levels = letters[1:4]),
                  d = Sys.Date())
class(data$x) <- c("blah_class")

fwrite(data, file, yaml = TRUE)
x <- fread(file, yaml = TRUE)


class(x$x)
#> [1] "integer"
attr(x, "yaml_metadata")$schema$fields[[1]]$type
#> [1] "blah_class"
```

<sup>Created on 2020-02-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by MichaelChirico** [comment]
We haven't seen much interest in csvy TBH, I will leave it as KTLO and possibly eventually deprecate it. My sense is that users that actually require having schema in their files are just using a binary format that builds in this support more naturally.

Here's the references to `csvy` on CRAN as of now:

https://github.com/search?q=org%3Acran+lang%3AR+%2Ffread%5C%28%2F+%2Fyaml%5Cs*%3D%2F+-repo%3Acran%2Fdata.table&type=code

NB {mindicador} was removed from CRAN 4 years ago https://cran.r-project.org/web/packages/mindicador/index.html
